ta/pkcs11/src/object.c

```diff
@@ -409,3 +409,591 @@ enum pkcs11_rc entry_destroy_object(struct pkcs11_client *client,
+	 * Scan all objects (sessions and persistent ones) and set a list of
+	 * candidates that match caller attributes. First scan all current
+	 * session objects (that are visible to the session). Then scan all
+	 * remaining persistent object for which no session object handle was
+	 * published to the client.
+	 */
+
+	LIST_FOREACH(obj, &session->object_list, link) {
+		uint32_t *handles = NULL;
+
+		rc = check_access_attrs_against_token(session, obj->attributes);
+		if (rc)
+			continue;
+
+		if (!attributes_match_reference(obj->attributes, req_attrs))
+			continue;
+
+		handles = TEE_Realloc(find_ctx->handles,
+				      (find_ctx->count + 1) * sizeof(*handles));
+		if (!handles) {
+			rc = PKCS11_CKR_DEVICE_MEMORY;
+			goto out;
+		}
+		find_ctx->handles = handles;
+
+		*(find_ctx->handles + find_ctx->count) =
+			pkcs11_object2handle(obj, session);
+		find_ctx->count++;
+	}
+
+	/* Remaining handles are those not yet published by the session */
+	find_ctx->temp_start = find_ctx->count;
+
+	LIST_FOREACH(obj, &session->token->object_list, link) {
+		uint32_t obj_handle = 0;
+		uint32_t *handles = NULL;
+
+		/*
+		 * If there are no attributes specified, we return
+		 * every object
+		 */
+		if (req_attrs->attrs_count) {
+			rc = token_obj_matches_ref(req_attrs, obj);
+			if (rc == PKCS11_RV_NOT_FOUND)
+				continue;
+			if (rc != PKCS11_CKR_OK)
+				goto out;
+		}
+
+		rc = check_access_attrs_against_token(session, obj->attributes);
+		if (rc)
+			continue;
+
+		/* Object may not yet be published in the session */
+		obj_handle = pkcs11_object2handle(obj, session);
+		if (!obj_handle) {
+			obj_handle = handle_get(&session->object_handle_db,
+						obj);
+			if (!obj_handle) {
+				rc = PKCS11_CKR_DEVICE_MEMORY;
+				goto out;
+			}
+		}
+
+		handles = TEE_Realloc(find_ctx->handles,
+				      (find_ctx->count + 1) * sizeof(*handles));
+		if (!handles) {
+			rc = PKCS11_CKR_DEVICE_MEMORY;
+			goto out;
+		}
+
+		/* Store object handle for later publishing */
+		find_ctx->handles = handles;
+		*(handles + find_ctx->count) = obj_handle;
+		find_ctx->count++;
+	}
+
+	if (rc == PKCS11_RV_NOT_FOUND)
+		rc = PKCS11_CKR_OK;
+
+	/* Save target attributes to search (if needed later) */
+	find_ctx->attributes = req_attrs;
+	req_attrs = NULL;
+	session->find_ctx = find_ctx;
+	find_ctx = NULL;
+	rc = PKCS11_CKR_OK;
+
+out:
+	TEE_Free(req_attrs);
+	TEE_Free(template);
+	release_find_obj_context(session, find_ctx);
+
+	return rc;
+}
+
+enum pkcs11_rc entry_find_objects(struct pkcs11_client *client,
+				  uint32_t ptypes, TEE_Param *params)
+{
+	const uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INOUT,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_MEMREF_OUTPUT,
+						TEE_PARAM_TYPE_NONE);
+	TEE_Param *ctrl = params;
+	TEE_Param *out = params + 2;
+	enum pkcs11_rc rc = PKCS11_CKR_GENERAL_ERROR;
+	struct serialargs ctrlargs = { };
+	struct pkcs11_session *session = NULL;
+	struct pkcs11_find_objects *ctx = NULL;
+	char *out_handles = NULL;
```

Comment:
`uint8_t *` to be more specific that it is raw serialized data?

Reply:
`char *` and `uint8_t *` are both fine to me.

Reply:
Changing it to uint8_t* for clarity